### PR TITLE
Update DistributedData Delete API to include optional request context. #20140

### PR DESCRIPTION
**Pull request by inakov**
- Included optional request context to Delete API.
- Added new message for Subscribers when key is removed.

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @inakov,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by patriknw** [comment]
Refs #20140

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 1]
```diff
@@ -350,6 +351,10 @@ object Replicator {
     def dataValue: A = data
   }
 
+  final case class Deleted[A <: ReplicatedData](key: Key[A]) extends NoSerializationVerificationNeeded {
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: Add a new section for Distributed Data in akka-docs/rst/project/migration-guide-2.4.x-2.5.x.rst and mention this change.

**Review by patriknw — COMMENTED** [review]
great, just a note in the migration guide

File: project/MiMa.scala [hunk 2]
```diff
@@ -663,7 +663,25 @@ object MiMa extends AutoPlugin {
 
         // #21894 Programmatic configuration of the ActorSystem
         ProblemFilters.exclude[DirectMissingMethodProblem]("akka.actor.ActorSystemImpl.this")
-      ) ++ bcIssuesBetween24and25)
+      ) ++ bcIssuesBetween24and25),
+      "2.4.15" -> Seq(
```
Comment (patriknw) [inline project/MiMa.scala]: 2.4.16

**Comment by inakov** [comment]
@patriknw, I updated the migration guide and changed the version in MiMa.scala.
Anything else?

**Review by patriknw — APPROVED** [review]
LGTM, that's great. Another reviewer and then we can merge.

File: project/MiMa.scala [hunk 3]
```diff
@@ -663,7 +663,25 @@ object MiMa extends AutoPlugin {
...
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#ReplicationDeleteFailure.copy"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#ReplicationDeleteFailure.this"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#DeleteSuccess.copy"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#DeleteSuccess.this"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#Delete.apply"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#DataDeleted.apply"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#DataDeleted.copy"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#DataDeleted.this"),
+        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.ddata.Replicator#Delete.copy")
+      )
```
Comment (patriknw) [inline project/MiMa.scala]: @inakov sorry, I was wrong. Revert this and place these in the section bcIssuesBetween24and25 instead.

Please rebase also, merge conflict

**Comment by patriknw** [comment]
@ivan-lorenz Please rebase and squash to one commit and then we can merge

**Comment by inakov** [comment]
I'm on it! :)

**Comment by inakov** [comment]
Hi @patriknw, sorry but I messed up my branch history, I'll be more careful next time ;( .I created one new PR with branch that is going to be easier for merge. I'm going to do the same thing for the other. Is that OK? Can I close this PR? This is the new PR #22142

**Pull request closed without merging**